### DISCUSSION TITLE
Information about Blizzard Spell Ticks for use with APL

### DISCUSSION THREAD

**Discussion by flollef** [simulationcraft/simc#6305] [Q&A] [answered]
Is there currently a way to get information about when blizzard is ticking (and when it was casted), how many ticks left, etc for use with the action list? The commands for dots do not seem to work with blizzard.

**Comment by vituscze** [comment, marked as answer]
See https://github.com/simulationcraft/simc/wiki/Mages#ground-aoe. Other information is not available to the APL.

**Comment by flollef** [comment]
Is there any possibility that more information about blizzard could me made available to the APL? Blizzard damage controls the cooldown of frozen orb, and any attempt to predict when frozen orb will come off cooldown relies on information about when blizzard will deal its damage. Knowing when orb is coming off cooldown is necessary to optimize situations where the player has the choice of either casting or waiting for orb.

It is definitely possible to estimate when the blizzard ticks are occurring from the information available now, but it is much more complicated and likely not as reliable as direct access to the information. Note that blizzard ticks in game appear to follow a slightly irregular pattern in their timing as interpreted by the combat log.

**Reply by vituscze** [reply]
Probably not, we would have to estimate that information in simc too (tick time is dynamic and we can't reliably predict it).

Blizzard ticks are very regular. If you have a log that shows otherwise, please share it with us.

**Comment by flollef** [comment]
Perhaps tracking the time of last tick (rather than future ticks) would be simpler to implement?

With respect to the irregularity of blizzard ticks with 0% haste a sample of combat log (time extracted via a weak aura using the
select(1, ...) expression on the combat log event) shows time between ticks as

1.067
1.018
1.000
0.951
0.968
1.000
1.068
1.001

With 9 total ticks, the total time is 8.073 seconds 

This could be a latency issue that applies to dots in general as I haven't done extensive testing on their in-game behavior.

**Reply by vituscze** [reply]
This seems fairly normal. The log is accurate, but not very precise.

As for the other question: is `ceil(ground_aoe.blizzard.remains%spell_haste)*0.5*active_enemies` really that far off that a custom expression is necessary?

**Comment by flollef** [comment]
Some imprecision is fine. The concern I have is related to situations where the time is right on or very close to the tick time, and the APL thinks the tick has happened and it hasn't (or vice versa). Any problem with this sort of exception could be avoided with time of last tick information.

**Reply by vituscze** [reply]
Yes, I understand that. I'm not convinced that such precision is necessary, though. If you can demonstrate that such detailed tracking of Blizzard's cooldown reduction leads to a DPS increase, I will consider adding better expression support.

**Comment by flollef** [comment]
The tricky part is testing any potential dps gain requires the precise cooldown tracking of frozen orb. 

However, delaying blizzard while it is off cd can be tested and shares some characteristics with delaying orb while it is off cd, in that both cause missed cd reduction ticks of blizzard on orb. 

A bit of preliminary testing with delaying blizzard in the following scenario ->

Necrolord Covenant, gear all set at 278 intellect level, crit set to 992, haste set to 993, vers set to 464  Death's Fathom + Cold Front Legendary + IQD all equipped

Action list is essentially 
Standard Cooldowns
Frozen Orb
Blizzard 
Frostbolt

This is not 100% optimal as Frostbolt's value relative to other spells is at a minimum dependent on the number of targets, how many frostbolts have been cast and the amount of time left in the pull (or the time left on deathborne if that is the constraining factor), but is the simplest way to test the effects of delaying blizzard for frostbolt casts or vice versa.

In an 8 target scenario lasting 40 seconds and no lust, a fairly typical M+ pull, waiting to cast blizzard if the remaining cooldown is below a static threshold rather than just casting frostbolt if orb or blizzard are unavailable can result in a 2-3% increase in damage.

This is a fairly small increase but does show some potential for optimizing the timing of blizzard casts in certain situations, even without attempting to account for the changing value of frostbolt during an encounter.

**Comment by flollef** [comment]
With the nerfs out this may all be a moot point
